include/seqan3/range/view/async_buffer.hpp

```diff
@@ -0,0 +1,499 @@
+{
+    //!\brief The sentinel type to compare to.
+    using sentinel_type = std::ranges::default_sentinel_t;
+
+    //!\brief The pointer to the associated view.
+    contrib::fixed_buffer_queue<value_type_t<urng_t>> * buffer_ptr = nullptr;
+
+    //!\brief The cached value this iterator holds.
+    value_type_t<urng_t> cached_value;
+
+    //!\brief Whether this iterator is at end (the buffer is empty and closed).
+    bool at_end = false;
+
+public:
+
+    /*!\name Associated types
+    * \{
+    */
+    //!\brief Difference type.
+    using difference_type   = difference_type_t<urng_iterator_type>;
+    //!\brief Value type.
+    using value_type        = value_type_t<urng_iterator_type>;
+    //!\brief Pointer type.
+    using pointer           = value_type *;
+    //!\brief Reference type.
+    using reference         = value_type &;
+    //!\brief Iterator category.
+    using iterator_category = std::input_iterator_tag;
+    //!\}
+
+    /*!\name Construction, destruction and assignment
+     * \{
+     */
+    async_buffer_iterator()                                              noexcept = default; //!< Defaulted.
+    //TODO: delete:
+    async_buffer_iterator(async_buffer_iterator const & rhs)             noexcept = default; //!< Defaulted.
+    async_buffer_iterator(async_buffer_iterator && rhs)                  noexcept = default; //!< Defaulted.
+    //TODO: delete:
+    async_buffer_iterator & operator=(async_buffer_iterator const & rhs) noexcept = default; //!< Defaulted.
+    async_buffer_iterator & operator=(async_buffer_iterator && rhs)      noexcept = default; //!< Defaulted.
+    ~async_buffer_iterator()                                             noexcept = default; //!< Defaulted.
+
+    //!\brief Constructing from the underlying seqan3::async_buffer_view.
+    async_buffer_iterator(contrib::fixed_buffer_queue<value_type_t<urng_t>> & buffer) noexcept :
+        buffer_ptr{&buffer}
+    {
+        ++(*this); // cache first value
+    }
+    //!\}
+
+    /*!\name Access operations
+     * \{
+     */
+    //!\brief Dereferences the cached iterator.
+    reference operator*() noexcept
+    {
+        return cached_value;
+    }
+
+    //!\brief Dereferences the cached iterator.
+    value_type const & operator*() const noexcept
+    {
+        return cached_value;
+    }
+
+    //!\brief Returns pointer to the pointed-to object.
+    pointer operator->() const noexcept
+    {
+        return std::addressof(cached_value);
+    }
+    //!\}
+
+    /*!\name Iterator operations
+     * \{
+     */
+    //!\brief Pre-increment.
+    async_buffer_iterator & operator++() noexcept
+    {
+        if (at_end) // TODO unlikely
+            return *this;
+
+        assert(buffer_ptr != nullptr);
+
+        if (buffer_ptr->wait_pop(cached_value) == contrib::queue_op_status::closed)
+            at_end = true;
+
+        return *this;
+    }
+
+    //!\brief Post-increment.
+    void operator++(int) noexcept
+    {
+        ++(*this);
```

Comment:
so this iterator becomes no Cpp17Iterator. In case this matters.

Reply:
It's not required: https://en.cppreference.com/w/cpp/named_req/InputIterator

Reply:
> value_type x = *i;
++i;
return x;

> *i++ -> convertible to value_type

Reply:
And further down:
>   typename std::common_reference_t<decltype(*i++)&&, typename std::readable_traits<I>::value_type&>;

Reply:
You are correct, but we wouldn't be able to satisfy "equivalent expression" in any case, right?

Reply:
so we leave it as no CPP17Iterator. You might want to add then a sentence to the view description that the returned iterator does not model Cpp17InputIterator. After this we can merge it.